rubullet/src/types.rs

```diff
@@ -0,0 +1,1128 @@
+    pub world_pose: Isometry3<f64>,
+    /// local offset of the inertial frame (center of mass) express in the URDF link frame
+    pub local_inertial_pose: Isometry3<f64>,
+    /// world pose of the URDF link frame
+    pub world_link_frame_pose: Isometry3<f64>,
+    ///Cartesian world linear  velocity. Only valid when ACTUAL_STATE_COMPUTE_LINKVELOCITY is set (b3RequestActualStateCommandComputeLinkVelocity)
+    pub world_linear_velocity: [f64; 3],
+    ///Cartesian world angular velocity. Only valid when ACTUAL_STATE_COMPUTE_LINKVELOCITY is set (b3RequestActualStateCommandComputeLinkVelocity)
+    pub world_angular_velocity: [f64; 3],
+}
+impl From<b3LinkState> for LinkState {
+    fn from(b3: b3LinkState) -> Self {
+        let b3LinkState {
+            m_world_position,
+            m_world_orientation,
+            m_local_inertial_position,
+            m_local_inertial_orientation,
+            m_world_link_frame_position,
+            m_world_link_frame_orientation,
+            m_world_linear_velocity,
+            m_world_angular_velocity,
+            m_world_aabb_min: _,
+            m_world_aabb_max: _,
+        } = b3;
+        LinkState {
+            world_pose: position_orientation_to_isometry(m_world_position, m_world_orientation),
+            local_inertial_pose: position_orientation_to_isometry(
+                m_local_inertial_position,
+                m_local_inertial_orientation,
+            ),
+            world_link_frame_pose: position_orientation_to_isometry(
+                m_world_link_frame_position,
+                m_world_link_frame_orientation,
+            ),
+            world_linear_velocity: m_world_linear_velocity,
+            world_angular_velocity: m_world_angular_velocity,
+        }
+    }
+}
+
+fn position_orientation_to_isometry(position: [f64; 3], orientation: [f64; 4]) -> Isometry3<f64> {
+    Isometry3::<f64>::from_parts(
+        Translation3::from(Vector3::from_column_slice(&position)),
+        UnitQuaternion::from_quaternion(Quaternion::from_parts(
+            orientation[3],
+            Vector3::from_column_slice(&orientation[0..3]),
+        )),
+    )
+}
+fn combined_position_orientation_array_to_isometry(combined: [f64; 7]) -> Isometry3<f64> {
+    let position = [combined[0], combined[1], combined[2]];
+    let orientation = [combined[3], combined[4], combined[5], combined[6]];
+    position_orientation_to_isometry(position, orientation)
+}
+
+/// VisualShape options are for the [create_visual_shape](`crate::PhysicsClient::create_visual_shape`)
+/// function to specify additional options like the color.
+pub struct VisualShapeOptions {
+    /// offset of the shape with respect to the link frame
+    pub frame_offset: Isometry3<f64>,
+    /// color components for red, green, blue and alpha, each in range [0,1]
+    pub rgba_colors: [f64; 4],
+    /// specular reflection color, red, green, blue components in range [0,1]
+    pub specular_colors: [f64; 3],
```

Comment:
puh. thats a lot of typing for my taste ```Rgba<f64>([1.;4])``` . If it could assert that every number is in the range [0,1] it would be great. Or if the image crate could convert HSV to RGB so that you could easily generate different looking colors.

Without that it just creates verbosity imo.

Reply:
You'll never actually have to type that. In the function signature, it's just `Rgba<f64>` and at the call site it's `Rga([r, g, b, a])`. You're right about it not adding much utility beyond the type, though, so if you don't want to make the change I won't fight it.

Reply:
Ok then I would say we leave it as it is (I converted the slices to arrays though) and wait until there comes a Color type that is widely used and can do a bit more